### ISSUE TITLE
"Login to update" lookup API

### ISSUE THREAD

**Issue by grigoryk** [mozilla/application-services#2878]
We're improving logic in A-C which is responsible for determining if an "incoming" (from GV, i.e. a user interaction with a login page) record should trigger an update or create operation. See https://github.com/mozilla-mobile/android-components/pull/6298

It makes sense (to me!) to consolidate this logic in the logins library. Currently, we have some heuristics in A-C, some in Rust, and it's quite hard to understand what exactly is going on. The current approach is also quite inefficient, and we're in a hot-path here for handling user input (i.e. all of this code runs while user is interacting with the login doorhangers).

Proposed shape of this new method is, roughly:
```
@Throws(misc invalid record exceptions)
fun lookupRecordToUpdate(incomingLogin: LoginRecord): LoginRecord?
```

- If this throws, we have a malformed incoming record, or a dupe.
- If this returns nothing, we know that `incomingLogin` will create a new record
- if this returns something, we know that `incomingLogin` should be used to update that record

On a-c side, we'll invoke this API, and depending on the response will update the UIs accordingly. In case of an update, the returned record from the store will be mutated w/ the fields changed in `incomingLogin`.

This new method can (probably) replace `ensureValid`.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1019)
┆Sprint: Backlog

**Comment by mnoorenberghe** [comment]
We talked about a replacement for the current API at All Hands that would return all of the usernames for the "site" so the username comparison could be done by the consuming app. Since we would need to look at storage to see if there is a login to update I think the API I describe would be better for a hot path.

**Comment by thomcc** [comment]
The API @mnoorenberghe describes is available https://github.com/mozilla/application-services/pull/2542, although I haven't had time to touch it in a while.

**Comment by grigoryk** [comment]
After chatting w/ @mnoorenberghe the proposed API in #2542 makes perfect sense to me, and should serve our needs better than what I describe here.

@thomcc other than tests it looks pretty close, can we land it?

**Comment by thomcc** [comment]
Fixed by #2542
